Fix #5361 - attempt at least, with js load after window complete

dnil opened the pull request:
This PR adds a functionality or fixes a bug.

- Fix #5361 - use js call to api, to chanjo2 extension, to set indicator after window load completion

![Screenshot 2025-03-31 at 13 46 43](https://github.com/user-attachments/assets/c18406fd-bff7-4090-a0d2-a1b5618b8f71)
![Screenshot 2025-03-31 at 13 43 20](https://github.com/user-attachments/assets/938fab45-0cfa-44ef-819d-ba8f0b30a7b4)
![Screenshot 2025-03-31 at 13 45 12](https://github.com/user-attachments/assets/a17799e9-5b63-41b0-a582-df37fb96e0fb)

This is about what I meant. Looks tentatively ok, but I'm a little worried about the test server load still with this. We will be spamming the chanjo2 service, and if d4tools is going to take some time each call.

Anyway, looks like the tests are also bugging with the latest and greatest in the lock file. Presumably some library has updated with some significance, but one thing at a time.

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>


**How to test**:
1. how to test it, possibly with real cases/data

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

northwestwitch commented:
> Looks like a good solution to me. I can't really say if it makes a difference because loading of the button is too fast on our stage server, so perhaps @Jakob37 could give it a try instead?
> 
> Could something be done for the icons? They don't look super clear I think:
> 
> <img alt="image" width="172" src="https://private-user-images.githubusercontent.com/28093618/428635594-f34f3afc-80a2-4e00-8816-5a316a0ebad7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0Mjc4NzcsIm5iZiI6MTc0MzQyNzU3NywicGF0aCI6Ii8yODA5MzYxOC80Mjg2MzU1OTQtZjM0ZjNhZmMtODBhMi00ZTAwLTg4MTYtNWEzMTZhMGViYWQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMzElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzMxVDEzMjYxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljZmVhNDEyOTkzNWViZjNjOTU4OTJjODE5NTYyODMxYTY3N2RmYTYyOGUwY2UwZDg5MDAxNzUzOTA5ODAwZDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.3enyX8v2o-zpyUsj4B55uF70TtWtmYhWEf5j4wQB3H0">

Perhaps just removing the second button and having just one green button that says "complete coverage" or a yellow one with "incomplete coverage"?

Jakob37 commented:
I'll give it a go!

Jakob37 commented:
![Animation](https://github.com/user-attachments/assets/889e9b9f-db36-4cb0-abe1-666efd8c4624)

Works fine! Async is a good choice here I think. Worked fine opening a bunch of variants as well.

It means a lot of data transfer though. Unsure whether an issue when server are under more load here. Could there be an institute-option to disable these async calculations?

Finally, no request, just a user question - What about a mini-spinner instead of a static question mark when waiting to signal things are happening?

Nice job 💪 👌

dnil commented:
![Screenshot 2025-04-23 at 16 07 50](https://github.com/user-attachments/assets/0a3c8044-6363-4b56-966d-5dee0925c0a3)
![Screenshot 2025-04-23 at 16 08 18](https://github.com/user-attachments/assets/3dcf7149-3fce-405c-9873-0d6a907a3284)


Ok, this should be ready for another check:
- added a button-looking span to get some more visible status color: perhaps we could merge that with the report button itself to save space at some point, but I didn't want to refactor that function as well now. This also looks a little better when the window is narrow, and the elements end up stacked over one another.
- added a fadein-fadeout pulse to the "Checking" question mark. It's kind of pretty, but with any luck no-one will ever notice it. 😄 

We could absolutely make this on demand, either by expecting a click to check, or by institute settings, but I feel we will need to sort any performance issues with the chanjo2-api regardless? The users may just as well click each variant coverage report, or multiple HPO term reports.

dnil commented:
Ok like this for you both? See anything odd or is all ok still?

northwestwitch commented on scout/server/blueprints/variant/templates/variant/components.html:
```diff
@@ -161,12 +161,8 @@
-          <span class="bg-warning fa-solid fa-triangle-exclamation" data-bs-toggle="tooltip" data-bs-placement="bottom" title="Note that Chanjo2 coverage is below 100% completeness."></span>
-        {% endif %}
+        {{ chanjo2_report_form(institute, case, gene.hgnc_symbol if gene.hgnc_symbol else gene.hgnc_id, 'overview', gene.hgnc_id, "Gene coverage " ~ (gene.hgnc_symbol if gene.hgnc_symbol else gene.hgnc_id), "btn btn-sm btn-secondary text-white") }} <!--chanjo2 genes overview -->
+        <span class="button complete-coverage-indicator btn-info text-white btn-sm" style="display:flex" id="coverage-indicator-{{gene.hgnc_id}}-button" data-bs-toggle="tooltip" data-bs-placement="bottom" data-loading-text="Checking completeness..." title="Checking completeness..."><span id="coverage-indicator-{{gene.hgnc_id}}-icon" class='blink_me fa-solid fa-circle-question'></span><span id="co ...
```
This works very nicely. About the looks of this button, but I noticed that sometimes it's aligned with the coverage button and sometimes it isn't:

<img width="583" alt="image" src="https://github.com/user-attachments/assets/46d3c9ad-11e3-4d5d-a2ae-6f7b4bc6d8e9" />

<img width="664" alt="image" src="https://github.com/user-attachments/assets/de745800-c0bd-4512-a77a-a635d575944b" />


Turning these 2 buttons into a button group could help here. Is it doable? If this is doable then you can also omit the name of the gene in the second button

Also the icon inside the button is misaligned with the text:

<img width="189" alt="image" src="https://github.com/user-attachments/assets/cb921656-c654-42f4-8d1b-0240dcf0dd3d" />

dnil replied:
Sure, lets try at least! It would be less repetitive, but slightly wider by default without the stacking.

northwestwitch reviewed (approved):
Very nice with the loading! I have a suggestion for a couple of improvements to the aesthetics of the button, check my comment.

Jakob37 commented:
I like the pulsing - it looks pleasing and clearly shows that something is going on 😄 

I agree with the styling comments already mentioned by Chiara. Nothing else from me.

dnil commented:
Ok, so, I moved the coverage buttons down to their own list item row to make the display a little less crowded. The local obs column is usually larger anyway, so it's not like it is wasting vertical space. I think I managed to coerce all spans to be proper flex style, perhaps overly so, but now they align pretty well. The aestethics when there are multiple genes for the same variant could be better, but the remainder will be a little complicated to fix with different length gene names.

Narrow display, two genes:
![Screenshot 2025-04-28 at 15 32 49](https://github.com/user-attachments/assets/b2f9c948-dba2-4512-9f41-c04d9694112a)

Wide display, two genes:
![Screenshot 2025-04-28 at 15 33 06](https://github.com/user-attachments/assets/8f10cd32-cc7d-4ba1-afbf-9bd3f3af01df)

One gene: 
![Screenshot 2025-04-28 at 15 33 49](https://github.com/user-attachments/assets/d9d50618-fb41-4cd8-9710-50b647b94e37)

I also moved the tooltips a little so we only get one for chanjo, one for chanjo2 coverage.

northwestwitch commented:
It looks very nice, thanks!
